### PR TITLE
fix waitUntil handling in pipeToNodeResponse

### LINKED ISSUE #52962: Vercel deployment uses stale data cache ignoring previous re-validations with revalidateTag
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #22 SMP Tue Jan 10 18:39:00 UTC 2023
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: 7.1.0
    Relevant Packages:
      next: 13.4.11-canary.0
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 4.9.5
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/next-52962-w2xxc5

### To Reproduce

1) Clone https://codesandbox.io/p/sandbox/next-52962-w2xxc5 into a GitHub repository
2) Deploy GitHub repository to Vercel
3) Open the deployment URL, make note of the timestamp on the page
4) Click the Refresh button (reload the page if the data doesn't refresh on its own)
5) Make note of the timestamp on the page (it is expected to be different from the one in step 3)
6) Make any code change and commit to GitHub triggering new deployment on Vercel
7) Open the deployment URL, make note of the timestamp on the page - it's the same as in step 3 

The deployment in step 6 used stale data cache from deployment in step 2 completely ignoring data re-validation in step 4.

### Describe the Bug

Vercel deployments re-use stale data cached from previous deployments completely ignoring data re-validation performed by `revalidateTag` since previous deployment.

### Expected Behavior

New deployments must either fetch new data or use the results of the last fetch request, not the data from previous deployment which could have happened days ago.

### Which browser are you using? (if relevant)

_No response_

### How are you deploying your application? (if relevant)

_No response_

### LINKED ISSUE #57632: fetch revalidation not working with Nextjs 14.0.0
### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/blissful-boyd-mh27hy

#### Home Page
```tsx
import Link from "next/link";

const Home = () => {
  return (
    <div>
      <Link href="/time" className="border-2 p-2 mx-auto">
        Time
      </Link>
    </div>
  );
};

export default Home;
```

#### TimePage
```tsx
const Time = async () => {
  const res = await fetch("http://localhost:3000/api", {
    next: { revalidate: 2 },
  });
  const data = res.json();

  return <div>{data}</div>;
};

export default Time;
```

#### GET Route.
```tsx
import { NextResponse } from "next/server";
export const GET = () => {
  return NextResponse.json(Date.now());
};

export const dynamic = "force-dynamic";
```
Step 1 : Navigate to `Time` using link on home page.
Step 2 : Navigate back to `Home`.
Step 3 : wait for more than 2 seconds.
Step 4 : Again navigate to `Time`.
You see that The Cache is not changed even after completion of 2 seconds which was mentiond in `fetch` of `Time`. 

https://github.com/vercel/next.js/assets/87939462/0cdda015-7599-415c-bcf4-ad634e1d1063

### Current vs. Expected behavior

The cache is not updated which should be updated when mentioned revalidation time has completed.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.0.0: Fri Sep 15 14:41:34 PDT 2023; root:xnu-10002.1.13~1/RELEASE_ARM64_T8103
Binaries:
  Node: 19.4.0
  npm: 9.6.7
  Yarn: 1.22.17
  pnpm: N/A
Relevant Packages:
  next: 14.0.0
  eslint-config-next: 14.0.0
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ztanner**
### What?
Calls to `revalidateTag` when using the data cache would sometimes be ignored. This would result in stale data being displayed even after manually triggering a revalidation. 

### Why?
Revalidation calls are pushed onto a `pendingRevalidates` array and, in the case of a route handler, are awaited [here](https://github.com/vercel/next.js/blob/8b11264ea95c9ed2f7dfd1e8d66a36b57ebc0da1/packages/next/src/server/send-response.ts#L53-L55).  

Previously this took place as part of the former `pipeReadable` implementation ([ref](https://github.com/vercel/next.js/blob/57bb52d37dd43fd669c5c07296daba19b6a12917/packages/next/src/server/send-response.ts#L49)) which awaited the promise before calling res.end ([ref](https://github.com/vercel/next.js/blob/57bb52d37dd43fd669c5c07296daba19b6a12917/packages/next/src/server/pipe-readable.ts#L99-L105)). There was a subtle change in behavior here in a recent refactor that seems to cause a race between awaiting the promise and ending the response, which results in this unpredictable behavior.

### How?
This applies similar logic that used to be in the `pipeReadable` function to the new `pipeToNodeResponse` function. Namely, we plumb `waitUntil` into the writeable stream close handler and await it before calling `res.end()`.

This is a very difficult edge case to test, and as such resorted to rigorous manual testing when deployed to an environment that uses the data cache. 

Fixes #52962
Fixes #57632

### CODE REVIEW

**Comment by ztanner** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #58744** <a href="https://app.graphite.dev/github/pr/vercel/next.js/58744?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/58744?utm_source=stack-comment).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| buildDuration | 10.4s | 10.6s | ⚠️ +195ms |
| buildDurationCached | 6.1s | 6s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +7.45 kB |
| nextStartRea..uration (ms) | 401ms | 401ms | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| 199-HASH.js gzip | 28.7 kB | 28.7 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| 494.HASH.js gzip | 180 B | 181 B | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 240 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.7 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 100 kB | 100 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 501 B | 503 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | ✓ |
| edge-ssr-HASH.js gzip | 253 B | 255 B | N/A |
| head-HASH.js gzip | 348 B | 347 B | N/A |
| hooks-HASH.js gzip | 369 B | 368 B | N/A |
| image-HASH.js gzip | 4.28 kB | 4.27 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.61 kB | 2.6 kB | N/A |
| routerDirect..HASH.js gzip | 311 B | 311 B | ✓ |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.17 kB | 3.17 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 483 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| index.html gzip | 527 B | 528 B | N/A |
| link.html gzip | 538 B | 542 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| edge-ssr.js gzip | 92.4 kB | 92.4 kB | N/A |
| page.js gzip | 145 kB | 145 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 628 B | 626 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 24.8 kB | 24.8 kB | ✓ |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 26.7 kB | 26.7 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 11-21-fix_waitUntil_handling_in_pipeToNodeResponse | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 167 kB | 167 kB | N/A |
| app-page-exp..prod.js gzip | 93.2 kB | 93.2 kB | N/A |
| app-page-tur..prod.js gzip | 94 kB | 94 kB | N/A |
| app-page-tur..prod.js gzip | 88.5 kB | 88.6 kB | N/A |
| app-page.run...dev.js gzip | 137 kB | 137 kB | N/A |
| app-page.run..prod.js gzip | 87.9 kB | 87.9 kB | N/A |
| app-route-ex...dev.js gzip | 23.8 kB | 23.8 kB | ✓ |
| app-route-ex..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route-tu..prod.js gzip | 16.5 kB | 16.5 kB | ✓ |
| app-route-tu..prod.js gzip | 16 kB | 16 kB | ✓ |
| app-route.ru...dev.js gzip | 23.2 kB | 23.2 kB | ✓ |
| app-route.ru..prod.js gzip | 16 kB | 16 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.8 kB | 21.8 kB | N/A |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.8 kB | 21.8 kB | N/A |
| server.runti..prod.js gzip | 49.1 kB | 49.1 kB | N/A |
| Overall change | 163 kB | 163 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages-turbo...time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>pages.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>server.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 684a41dd96989b045af01a074506423dc93d5266

File: packages/next/src/server/pipe-readable.ts [hunk 1]
```diff
@@ -86,7 +93,8 @@ function createWriterFromResponse(
 
 export async function pipeToNodeResponse(
   readable: ReadableStream<Uint8Array>,
-  res: ServerResponse
+  res: ServerResponse,
+  waitUntilForEnd?: Promise<void>
```
Comment (wyattjoh) [inline packages/next/src/server/pipe-readable.ts:97]: This should probably be `Promise<unknown>`

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7597706268a98f74ee55c075f948ebd0ce01e6e0

`pnpm test-dev test/e2e/app-dir/app-compilation/index.test.ts` 
- app dir > HMR > should not cause error when removing loading.js

<details>
<summary>Expand output</summary>

  ● app dir › HMR › should not cause error when removing loading.js

    TIMED OUT: hello from new page

    hello from slow page

    undefined

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at Object.<anonymous> (e2e/app-dir/app-compilation/index.test.ts:44:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by wyattjoh — APPROVED** [review]

**Review by ijjk — APPROVED** [review]

**Review by timneutkens — APPROVED** [review]

**Review by feedthejim — APPROVED** [review]
